### PR TITLE
New role for spoke cluster management operations (detach, attach and clean GitOps)

### PR DESCRIPTION
**Pull request by ramperher**
New roles to:

- Detach a spoke cluster from a given hub cluster / Attach a spoke cluster to a given hub cluster
- Delete GitOps configuration in a hub cluster (to detach a spoke cluster)

Test-Hints: no-check

- [x] Remove GitOps resources + detach cluster - https://www.distributed-ci.io/jobs/e1b632b8-232b-42ec-a681-fbaea312856d/jobStates
- [x] Attach cluster - https://www.distributed-ci.io/jobs/5029bc94-20a6-4c50-9e9d-e446f88291d9/jobStates
- [x] Launch the complete pipeline (ACM hub, ZTP spoke, remove GitOps resources and detach cluster, then attach cluster) - https://www.distributed-ci.io/jobs?limit=20&offset=0&sort=-created_at&where=tags:pipeline-id:vpool.8720,state:active
- [x] Tested cluster detachment + attachment again after refactoring this change: https://github.com/redhatci/ansible-collection-redhatci-ocp/pull/366#issuecomment-2216810033

### CODE REVIEW

**Comment by tonyskapunk** [comment]
Since acm spoke cluster roles are related, please consider having a single role for them defining an action (attach, detach) for them

**Comment by ramperher** [comment]
> Since acm spoke cluster roles are related, please consider having a single role for them defining an action (attach, detach) for them

Isn't it better to have them separeted in different roles? I mean, the required variables for each role are slightly different, and they represent a really different action. I'll ask the team's opinion about this before making any change, but for me it makes sense to diferentiate between both roles.

File: roles/acm_attach_spoke_cluster/tasks/main.yaml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+- name: Create autoimport secret
+  block:
+    - name: Create temporary template directory
+      ansible.builtin.tempfile:
+        state: directory
+        suffix: build
+      register: _aasc_temp_dir
+
+    - name: Get kubeconfig content
+      ansible.builtin.slurp:
```
Comment (reihl) [inline roles/acm_attach_spoke_cluster/tasks/main.yaml:52]: set_fact can be simpler:
`ansible.builtin.set_fact:
     _assc_kubeconfig_content: "{{file_contents: "{{ lookup('file', aasc_cluster_kubeconfig_path) }}"}}"`

Comment (sshnaidm) [reply roles/acm_attach_spoke_cluster/tasks/main.yaml:52]: the lookup plugin will work only when it's localhost, slurp is more flexible and supports any host

File: roles/acm_attach_spoke_cluster/tasks/main.yaml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+    - name: Template file
+      ansible.builtin.template:
+        src: autoimport_secret.yml.j2
+        dest: "{{ _aasc_temp_dir.path }}/autoimport_secret.yml"
+        mode: '0644'
+      no_log: true
+
+    - name: Apply Secret to attach spoke cluster to hub cluster
+      community.kubernetes.k8s:
```
[the remark this answers is not in the document]
Comment (ramperher) [reply roles/acm_attach_spoke_cluster/tasks/main.yaml:70]: I've made this change, thanks for reporting

File: roles/acm_attach_spoke_cluster/tasks/main.yaml (hunk 1 above)
Comment (ramperher) [reply roles/acm_attach_spoke_cluster/tasks/main.yaml:52]: As @sshnaidm said, I'd prefer to use `slurp` because of its flexibility

**Comment by ramperher** [comment]
Individual roles are working separately, now I'm testing the whole chain in a single dci-pipeline call (create a spoke cluster, detach it and re-attach it)

**Comment by ramperher** [comment]
[Complete pipeline](https://www.distributed-ci.io/jobs?limit=20&offset=0&sort=-created_at&where=tags:pipeline-id:vpool.8720,state:active) is working, this is ready for review.

File: roles/acm_attach_spoke_cluster/README.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+Name                         | Type   | Required | Default                                            | Description
+---------------------------- | ------ | -------- | -------------------------------------------------- | ------------------------------------------------------------
+aasc_cluster_kubeconfig_path | string | yes      | -                                                  | Path to spoke cluster's kubeconfig file
+aasc_cluster_name            | string | yes      | -                                                  | Cluster name, used for identifying the ManagedCluster and the Namespace
+
+License
+-------
+
+GNU GENERAL PUBLIC LICENSE version 3
```
Comment (tonyskapunk) [inline roles/acm_attach_spoke_cluster/README.md:30]: Why GPLv3?

File: roles/acm_detach_spoke_cluster/README.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+--------------
+
+Name                        | Type   | Required | Default                                            | Description
+--------------------------- | ------ | -------- | -------------------------------------------------- | -------------------------------------------------------------
+adsc_cluster_name           | string | yes      | -                                                  | Cluster name, used for identifying the ManagedCluster and the Namespace
+
+License
+-------
+
+GNU GENERAL PUBLIC LICENSE version 3
```
Comment (tonyskapunk) [inline roles/acm_detach_spoke_cluster/README.md:23]: GPL?

File: roles/acm_detach_spoke_cluster/README.md
Comment (tonyskapunk) [inline roles/acm_detach_spoke_cluster/README.md:1]: Please include usage examples

File: roles/acm_attach_spoke_cluster/README.md
Comment (tonyskapunk) [inline roles/acm_attach_spoke_cluster/README.md:1]: Please include usage examples

File: roles/remove_ztp_gitops_resources/README.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+rzgr_policies_namespace     | string | no       | policies-sub                                       | Namespace for the policy generator template resources. It can not be the sabe as the clusters namespace.
+rzgr_extra_namespaces       | list   | no       | ["ztp-common", "ztp-group", "ztp-site"]            | Extra namespaces that are created for GitOps policy creation by default.
+rzgr_cluster_role_bindings  | list   | no       | ["gitops-policy", "gitops-cluster"]                | ClusterRoleBindings created for the deployment of GitOps Policy and SiteConfig.
+rzgr_private_repo_secret    | string | no       | private-repo                                       | Secret that will hold the private repo credentials.
+rzgr_argo_cd_known_host_cm  | string | no       | argocd-ssh-known-hosts-cm                          | ConfigMap that will save the ArgoCD SSH known hosts.
+
+License
+-------
+
+GNU GENERAL PUBLIC LICENSE version 3
```
Comment (tonyskapunk) [inline roles/remove_ztp_gitops_resources/README.md:35]: GPL?

File: roles/remove_ztp_gitops_resources/README.md
Comment (tonyskapunk) [inline roles/remove_ztp_gitops_resources/README.md:1]: Please include usage examples

**Review by tonyskapunk — COMMENTED** [review]
> I'll ask the team's opinion about this before making any change, but for me it makes sense to diferentiate between both roles.

Any word on this?

Aren't these two roles part of the management of a spoke cluster? i.e. would it be a use case to detach a spoke without attaching it later? Or, attaching a spoke cluster without detaching it first? Even if one or the other could be independent, both are part of the mgmt of the spoke into the Hub.


> I mean, the required variables for each role are slightly different, and they represent a really different action. 

Agreed in the fact that the inputs are different but that shouldn't be a problem, e.g.

```yaml
- name: Attach
  include_role: spoke_mgmt
  vars:
    sm_action: attach
    sm_kubeconfig_path: /path/to/spoke
    sm_name: spoke-name

- name: Detach
  include_role: spoke_mgmt
  vars:
    sm_action: detach
    sm_name: spoke-name
```

And the validation is made depending on the action to perform.

File: roles/acm_attach_spoke_cluster/README.md
Comment (ramperher) [reply roles/acm_attach_spoke_cluster/README.md:1]: done

File: roles/remove_ztp_gitops_resources/README.md
Comment (ramperher) [reply roles/remove_ztp_gitops_resources/README.md:1]: done

File: roles/acm_attach_spoke_cluster/README.md (hunk 3 above)
Comment (ramperher) [reply roles/acm_attach_spoke_cluster/README.md:30]: I copied the role template from this another one: https://github.com/redhatci/ansible-collection-redhatci-ocp/tree/main/roles/configure_ztp_gitops_apps#license, so if this needs to be fixed, we need to check other roles too.

I'll remove from here for the moment.

File: roles/acm_detach_spoke_cluster/README.md (hunk 4 above)
Comment (ramperher) [reply roles/acm_detach_spoke_cluster/README.md:23]: same

File: roles/remove_ztp_gitops_resources/README.md (hunk 5 above)
Comment (ramperher) [reply roles/remove_ztp_gitops_resources/README.md:35]: same

**Comment by ramperher** [comment]
> > I'll ask the team's opinion about this before making any change, but for me it makes sense to diferentiate between both roles.
> 
> Any word on this?
> 
> Aren't these two roles part of the management of a spoke cluster? i.e. would it be a use case to detach a spoke without attaching it later? Or, attaching a spoke cluster without detaching it first? Even if one or the other could be independent, both are part of the mgmt of the spoke into the Hub.
> 
> > I mean, the required variables for each role are slightly different, and they represent a really different action.
> 
> Agreed in the fact that the inputs are different but that shouldn't be a problem, e.g.
> 
> ```yaml
> - name: Attach
>   include_role: spoke_mgmt
>   vars:
>     sm_action: attach
>     sm_kubeconfig_path: /path/to/spoke
>     sm_name: spoke-name
> 
> - name: Detach
>   include_role: spoke_mgmt
>   vars:
>     sm_action: detach
>     sm_name: spoke-name
> ```
> 
> And the validation is made depending on the action to perform.

I couldn't check with the team finally but I got your point and I've finally applied your proposed modifications, hope it's fine now.

**Comment by ramperher** [comment]
Sorry for the noise, I needed to validate again the refactor in the collections change, now it's ready again for review.

File: roles/acm_spoke_mgmt/README.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+
+### Role Variables
+
+Name                        | Type   | Required | Default                                            | Description
+--------------------------- | ------ | -------- | -------------------------------------------------- | -------------------------------------------------------------
+asm_cluster_name            | string | yes      | -                                                  | Cluster name, used for identifying the ManagedCluster and the Namespace
+
+### Example
+
+```
```
Comment (tonyskapunk) [inline roles/acm_spoke_mgmt/README.md:33]: nit:
```suggestion
```yaml
```

File: roles/acm_spoke_mgmt/README.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+Name                         | Type   | Required | Default                                            | Description
+---------------------------- | ------ | -------- | -------------------------------------------------- | ------------------------------------------------------------
+asm_cluster_kubeconfig_path  | string | yes      | -                                                  | Path to spoke cluster's kubeconfig file
+asm_cluster_name             | string | yes      | -                                                  | Cluster name, used for identifying the ManagedCluster and the Namespace
+
+
+### Example
+
+```
```
Comment (tonyskapunk) [inline roles/acm_spoke_mgmt/README.md:68]: nit:
```suggestion
```yaml
```

File: roles/acm_spoke_mgmt/tasks/main.yaml [hunk 8]
```diff
@@ -0,0 +1,23 @@
+---
+
+- name: Check if the selected action is allowed
+  vars:
+    asm_actions:
+      - detach
+      - attach
+  ansible.builtin.assert:
+    that:
+      - asm_action in asm_actions
```
Comment (tonyskapunk) [inline roles/acm_spoke_mgmt/tasks/main.yaml:10]: nit:
```suggestion
      - asm_action | lower in asm_actions
```

File: roles/remove_ztp_gitops_resources/tasks/main.yaml [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+- name: Delete SiteConfig and Policy Applications
+  community.kubernetes.k8s:
+    api_version: argoproj.io/v1alpha1
+    kind: Application
+    name: "{{ _rzgr_gitops_application }}"
+    namespace: openshift-gitops
+    state: absent
+    wait: true
+    wait_sleep: 5
+    wait_timeout: 120
```
Comment (tonyskapunk) [inline roles/remove_ztp_gitops_resources/tasks/main.yaml:15]: nit: These are the default `wait_sleep` and `wait_timeout` values, no need to set them unless they are different or will be modified by a variable

File: roles/acm_spoke_mgmt/tasks/attach.yaml [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+
+    - name: Save kubeconfig content in a variable
+      ansible.builtin.set_fact:
+        asm_kubeconfig: "{{ _asm_kubeconfig_content['content'] | b64decode }}"
+      no_log: true
+
+    - name: Apply Secret to attach spoke cluster to hub cluster
+      community.kubernetes.k8s:
+        state: present
+        definition: "{{ lookup('template', 'templates/autoimport_secret.yml.j2') | from_yaml }}"
```
Comment (tonyskapunk) [inline roles/acm_spoke_mgmt/tasks/attach.yaml:59]: nit:
```suggestion
        definition: "{{ lookup('ansible.builtin.template', 'autoimport_secret.yml.j2') | from_yaml }}"
```

**Review by tonyskapunk — COMMENTED** [review]
@ramperher thanks for all the work, I still have some minor comments, but is looking good.

File: roles/acm_spoke_mgmt/README.md (hunk 6 above)
Comment (ramperher) [reply roles/acm_spoke_mgmt/README.md:33]: ack

File: roles/acm_spoke_mgmt/README.md (hunk 7 above)
Comment (ramperher) [reply roles/acm_spoke_mgmt/README.md:68]: ack

File: roles/acm_spoke_mgmt/tasks/main.yaml (hunk 8 above)
Comment (ramperher) [reply roles/acm_spoke_mgmt/tasks/main.yaml:10]: ack

File: roles/remove_ztp_gitops_resources/tasks/main.yaml (hunk 9 above)
Comment (ramperher) [reply roles/remove_ztp_gitops_resources/tasks/main.yaml:15]: ack

File: roles/acm_spoke_mgmt/tasks/attach.yaml (hunk 10 above)
Comment (ramperher) [reply roles/acm_spoke_mgmt/tasks/attach.yaml:59]: ack

**Comment by ramperher** [comment]
> @ramperher thanks for all the work, I still have some minor comments, but is looking good.

@tonyskapunk I've applied all your feedback, thanks for the review!

**Review by tonyskapunk — APPROVED** [review]
Thanks for all the work in here, LGTM

**Review by betoredhat — APPROVED** [review]
Nice job! LGTM. Thanks Ramon.

**Pull request merged**